### DISCUSSION TITLE
How pattern matching works for basic numeric types

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/csharplang#8402] [Q&A]
Pattern matching checks the type. But in pattern matching, different types of numerical types can be mixed.
How should I understand this issue?
```csharp
float f = 1f;
Console.WriteLine(f is 1);
Console.WriteLine(f is 1f);
```
by the way.Both of these are true.but 1st generate CA1508 [false]

**Comment by HaloFour** [comment]
You're matching values, not types, and C# has built-in numeric type coercion.  1 and 1.0 are the same value as far as C# is concerned.

**Reply by zms9110750** [reply]
How about for this?
```csharp
object o = f;
Console.WriteLine(o is 1); //false
Console.WriteLine(o is 1f);//true
```

**Reply by HaloFour** [reply]
Since the operand is boxed and not a numeric type now it does require a type check and implicit numeric coercion is not applied.

**Reply by zms9110750** [reply]
So, if it's a value type, then 1, 1L, 1f, 1d, 1m are all the same?
If it is a reference type, will the type be checked first?
Where can I read more detailed rules for this judgment?

Also, I used INumbers for testing, but they don't match me now. Why is that
```csharp
INumber<float> n = f;
Console.WriteLine(n is 1); //CS0029
Console.WriteLine(n is 1f);//CS9060
```

**Reply by colejohnson66** [reply]
`INumber<float>` is actually a *boxed* float.

**Reply by HaloFour** [reply]
Numeric type coercion only works for the built-in primitive numeric types, not custom types nor does it consider implicit cast operators.

**Comment by huoyaoyuan** [comment]
Pattern matching is typed. They are effectively `(object)f is (int)1`, `(object)f is (float)1.0f`, and `(float)f is (float)1`.

Patterns are interpreted under the type of left hand side value.
